### DISCUSSION TITLE
Add ability to copy screenshots to clipboard

### DISCUSSION THREAD

**Discussion by uzervlad** [ppy/osu#16775] [Ideas] [answered]
Pretty self-explanatory, add a settings checkbox to automatically copy in-game screenshots to clipboard.

![osu_2022-02-04_13-53-09](https://user-images.githubusercontent.com/24654479/152517260-c7549f23-d81c-42b4-9e24-819bc18333b3.png)

**Comment by frenzibyte** [comment]
IMO a modifier (i.e. holding `Ctrl` when pressing `F12` / screenshot key) sounds better to be able to control when a screenshot should be copied, rather than a constant setting option.

**Reply by uzervlad** [reply]
True, dunno why I didn't think of that

**Reply by uzervlad** [reply]
Looks like pressing any modifier key prevents `GlobalAction.TakeScreenshot` from triggering, so
```cs
switch (e.Action)
{
    case GlobalAction.TakeScreenshot:
        shutter.Play();
        TakeScreenshotAsync(e.ControlPressed);
        return true;
}
```
doesn't work. Is there a possible workaround or will it have to be a separate keybind?

**Reply by frenzibyte** [reply]
Simplest solution would be to define a `GlobalAction.TakeScreenshotWithClipboard`, but would like to hear what the others say.

**Reply by bdach** [reply]
i disagree with a separate action. would rather have all screenshots available both on disk and in clipboard without any option to configure it, personally.

**Reply by peppy** [reply]
I'd agree with making the screenshot copied to clipboard a default behaviour until we have cases where a user complains about this (I'm not sure anyone would).

The one exception I can think of is mobile platforms, where accessing the clipboard potentially requires further permissions. Might be something to consider (although I think on iOS it is not an interactive permission request but a notification).

Keep in mind we will be bringing back the "Take screenshot and upload" which would copy a URL to the clipboard instead of the raw content. This would be a separate action.

**Comment by Stedoss** [comment]
This would be awesome! However, there doesn't appear to be a cross-platform API for the clipboard in .NET Core? Would this require a lot of weird interop with different environments?

**Reply by frenzibyte** [reply]
Correct, such support would start framework-side at `WindowsClipboard` / `MacOSClipboard` / etc., by implementing a `GetImage`/`SetImage` method.

Keep in mind it is not simple at all and I would not recommend it as a first contribution if you're planning to do that.

**Reply by Stedoss** [reply]
As someone who is scared to even PR small changes I will leave it to someone here who is much smarter than me lol, good luck to whoever picks this up! ^-^

**Reply by uzervlad** [reply]
I tried to implement copying images for `WindowsClipboard`, but I just couldn't find any info on the web on how to do it properly. Do you perhaps know where I could start doing my research on it? @frenzibyte

**Reply by frenzibyte** [reply]
I would suggest by having a look into the Windows Clipboard API reference (https://docs.microsoft.com/en-us/windows/win32/dataxchg/clipboard-reference).

**Reply by Stedoss** [reply]
I wonder if looking at how Powershell handles their `Get-Clipboard` and `Set-Clipboard` commands will also be helpful (albeit, only text): https://github.com/PowerShell/PowerShell/blob/7dc4587014bfa22919c933607bf564f0ba53db2e/src/Microsoft.PowerShell.Commands.Management/commands/management/Clipboard.cs

**Reply by uzervlad** [reply]
> I wonder if looking at how Powershell handles their `Get-Clipboard` and `Set-Clipboard` commands will also be helpful

I believe the only problem here is presenting the data to `SetClipboardData` correctly, so it probably won't be helpful

**Reply by uzervlad** [reply]
I just found out that the problem with my implementation was **probably** just not trimming first 14 bytes (bitmap file header). Will give it a try again tomorrow

**Reply by uzervlad** [reply]
So I confirmed it, it is now working, but I'm not sure I'll be able to do it for other platforms, especially since I have no way of testing it for Mac and iOS

https://user-images.githubusercontent.com/24654479/152633913-3727850d-7c04-47a1-87c0-211b0be7193a.mp4

**Reply by frenzibyte** [reply]
Begin by PR'ing your changes for the Windows platform, other platforms can have support separately.

**Comment by smoogipoo** [comment, marked as answer]
https://github.com/ppy/osu/pull/16907

**Comment by Speykious** [comment]
I just searched for this topic to see if it was already proposed, and turns out it had already been implemented since 2022.

So I wondered if it was actually working all this time for me (on Linux), and no, it just doesn't copy the image.

So I ran osu! on the latest branch and tried to see if the code was there and if it was working.
Turns out, it's there but not working! D:

![image](https://github.com/ppy/osu/assets/13885008/769ef147-58db-4534-b86f-1530ebc8a8d2)

It seems to be an osu!framework method, so I'll look into it...

**Reply by Speykious** [reply]
Ok, I found the issue. There just literally is no Linux implementation of the clipboard. `SDL2Clipboard.SetImage()` just returns `false`, `DesktopGameHost` doesn't override `SDL2GameHost.CreateClipboard` so it just instantiates an `SDL2Clipboard`, and same for `LinuxGameHost` as there is no `LinuxClipboard` as a counterpart to `WindowsClipboard` and `MacOSClipboard`.

**Reply by Speykious** [reply]
I have some implementation going on in https://github.com/ppy/osu-framework/pull/6151. It's not the best but at least it covers the majority of use cases on X11. I'm gonna look for improvements in the coming days

**Reply by Stedoss** [reply]
(bringing this out of PR to not muddy it up)
> To be completely fair, without considering the appimage factor, if any Linux user using X11 doesn't have `xclip` or has an `xclip` program that isn't the standard one every other program expects, it's pretty much guaranteed to be a user issue. If anything, this should at least tone down the lack of guarantee of `xclip` being the correct program, since it's just so ubiquitous in the Linux ecosystem.

Do note that this is not always the case. I have two different installs of Debian, one with just KDE X11 and one with KDE X11+Wayland and Cinnamon (X11). Both do not have `xclip` installed by default, so seems a little unfair to label it as a 'user issue'.

I don't think the answer is piping the file content into an external tool, I probably see it as having to call xlib/wayland directly. Big PITA but makes it a lot more portable.

**Reply by Speykious** [reply]
Yeah I agree. Calling into xlib and wayland directly is more portable. Problem is it requires that a library exists for this, or creating one.
Honestly if I don't find any such low-level solution, I'm gonna try making a native clipboard library that supports mimetypes, one that's at least gonna cover all these cases on Linux :v

**Reply by Speykious** [reply]
Also I don't want to make it seem like the implementation I provide there is the "answer". It's error-proof and very easily replaceable, so it can cover a majority of cases while waiting for a better implementation. I say that because I have no idea how long it takes to make a clipboard library that supports mime types both on X11 and Wayland.

**Reply by Speykious** [reply]
> I say that because I have no idea how long it takes to make a clipboard library that supports mime types both on X11 and Wayland.

Well given that I literally started making that, now I kinda got an idea. Might take at least 2 weeks of active coding there and that's probably underestimating :v

**Reply by bdach** [reply]
@Speykious consider contributing to https://github.com/jtanx/libclipboard or https://github.com/dacap/clip instead

**Reply by Susko3** [reply]
Or check if SDL3 rich clipboard works for you https://wiki.libsdl.org/SDL3/SDL_SetClipboardData.

We'll be updating to SDL3 at some point, so it's best not to reinvent the wheel.

**Reply by Speykious** [reply]
I did see that SDL3 seemed to have what it takes and was wondering if osu!framework ever planned to switch to it.

I guess I have to respond to @bdach here- as much as I don't like to admit it, I didn't even look at the C/C++ native libraries (aside from SDL). I'm mostly a Rust developer so I looked at the Rust solutions, and that's where the ones I've looked at don't support mime types or do it badly. Looking at them now though, it seems like `libclipboard` only supports text while `clip` seems to require libpng to paste images when setting the data directly is all we need. If I was a C/C++ dev I might've contributed to one of them instead.

On my side, I managed to get my clipboard solution working on X11. As it stands I can already find a way to make osu!framework use it as long as I create the C bindings, but I want to polish it a bit and also make the Wayland counterpart before that since it's important.

If SDL3 is right around the corner though, feel free to close my PR once it gets in... My clipboard solution will probably remain useful in the Rust ecosystem u_u

**Reply by Speykious** [reply]
Seems SDL3 is getting closer to a release...
https://github.com/libsdl-org/SDL/releases/tag/prerelease-3.1.0

**Reply by frenzibyte** [reply]
https://github.com/ppy/osu-framework/pull/6223

**Reply by bdach** [reply]
@frenzibyte that's not what that PR is really for though. and that pull only does anything on windows right now.
